### ISSUE TITLE
Repeat generation of aggregation keys

### ISSUE THREAD

**Issue by Stebalien** [filecoin-project/rust-fil-proofs#1479]
### Description

Currently, if multiple threads try to verify aggregate proofs with new parameters at the same time, we'll try to compute the keys multiple times (last one wins).

Ideally, we'd compute the keys once.

### Acceptance criteria

I can try to verify a thousand aggregate proofs at once without waiting a day or two (`(100s * 1000)/3600 = ~28 hours`).

### Risks + pitfalls

Extra complexity in the caches could introduce bugs. We also probably don't want to rely on nightly, which means either rolling our own "once" or using an external crate.

### Where to begin

Use something like https://doc.rust-lang.org/std/lazy/struct.SyncOnceCell.html (or https://crates.io/crates/conquer-once) or even just a mutex to define a per-key locked "cells" for every entry in our caches.

**Comment by Stebalien** [comment]
https://github.com/filecoin-project/rust-fil-proofs/blob/64a7c43abbb986bccb1d1e8532d711b269644994/filecoin-proofs/src/caches.rs#L40-L69

**Comment by cryptonemo** [comment]
cc: @nikkolasg This refers to the specialize per pow2 proof aggregation count

**Comment by cryptonemo** [comment]
Hrm, @Stebalien not sure that's the case, let me look, but that access is locked AFAIK

**Comment by cryptonemo** [comment]
Ah, I think I see -- the cache insertion is guarded, but not the generation

**Comment by Stebalien** [comment]
Yep. And we probably don't want to hold a _global_ lock here (maybe not terrible?).

**Comment by cryptonemo** [comment]
They aren't global locks, they are specific to the type (see lines above that method)

**Comment by cryptonemo** [comment]
@nikkolasg I still haven't gotten an answer on if there's any way to re-use the calculations across specialize calls (since we always calculate out to the max we support).  If so, we should chat about ways to do that before doing anything here since it could more generally speed things up.

**Comment by cryptonemo** [comment]
Noting also that the reason we've never run into this as an issue is because we have real parameter files that are never generated (so the generate path was previously not exercised ever since mainnet params were available).  While we don't generate fake SRS/IPP data either, we do take the time to do a compute intensive `specialize` call that displays this kind of breakage of cache usage.  Unfortunately this was apparently overlooked as the specialize perf was not supposed to be a significant issue.

**Comment by Stebalien** [comment]
> They aren't global locks, they are specific to the type (see lines above that method)

Global as in global for the entire cache (versus a lock for the specific key).

**Comment by cryptonemo** [comment]
I ran a similar bench, but of course I warmed it up by running once before spawning threads.  Is this an option in the short term if this is a significant issue for the first loading (i.e. verify once before parallelizing)?  There are 2 things that should be discussed on our side 1) Is it possible to lock just once no matter how many proofs are being aggregated since the long running part of the computation is always the same, or 2) Do we just leave it as is and come up with some hacky method of locking once per number of proofs being aggregated.

**Comment by cryptonemo** [comment]
If the answer to 1 is yes (cc: @nikkolasg), it could potentially be solved completely in bellman (where that data is re-used across calls), if 2, it could potentially be solved either in proofs or in lotus.  For 1, it could be too memory intensive or add another hacky cache file on disk, which will be problematic since like params it can be corrupted so it needs verification, etc.  For 2, it effectively serializes first time parallel verifies for each pow2 number of proofs, but keeps computations potentially high on slower machines for some time (i.e. the _original_ problem we had with specialize where it could take 100s several times over).  I personally prefer 1, with a memory only cache, but I don't know how much that would affect things (i.e. how big that would be).

**Comment by Stebalien** [comment]
> Is it possible to lock just once no matter how many proofs are being aggregated since the long running part of the computation is always the same

This should be fine. Under _normal_ circumstances, lotus will only verify one aggregate at a time.

However, we should compute these ahead-of-time, at compile time, or maybe even when initializing the repo (storing them on disk). We'd expect to compute each one at least once anyways. At the moment, a miner could get caught off guard having to compute multiple of these things on-demand and fail to produce a block in time.

Basically, we save nothing by computing them on-demand.

**Comment by Stebalien** [comment]
> For the cache itself I would recommend OnceCell which I usually use in cases like this, so that the cache becomes sth like

(ok, I swear I tried to find that _exact_ library and failed)
